File: javascript/card.js
/*
 * Copyright 2006 Stephen Duncan Jr
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at 
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

Card.ranks = ['Ace', 2, 3, 4, 5, 6, 7, 8, 9, 10, 'Jack', 'Queen', 'King'];
Card.suits = ['Clubs', 'Spades', 'Hearts', 'Diamonds'];

function Card(rank, suit) {
	this.rank = rank;
	this.suit = suit;
}

Card.prototype.toString = function() {
	return this.rank + " of " + this.suit;
}

Card.prototype.equal = function(other) {
	return this.rank == other.rank && this.suit == other.suit;
}

Card.prototype.compareTo = function(other) {
	var thisIndex = Card.ranks.indexOf(this.rank);
	var otherIndex = Card.ranks.indexOf(other.rank);

	if(thisIndex < otherIndex) {
		return -1;
	} else if(thisIndex > otherIndex) {
		return 1;
	} else {
		return 0;
	}
}

File: scala/src/main/scala/cards/Card.scala
package cards

sealed trait Suit
case object Clubs extends Suit
case object Spades extends Suit
case object Hearts extends Suit
case object Diamonds extends Suit

case class Card(rank: String, suit: Suit) extends Ordered[Card] {
  override def compare(that: Card): Int = Card.ranks.indexOf(this.rank) - Card.ranks.indexOf(that.rank)
  override def toString: String = s"$rank of $suit"
}

object Card {
  lazy val suits: Seq[Suit] = Seq(Clubs, Spades, Hearts, Diamonds)
  lazy val ranks: Seq[String] = "Ace" :: (1 to 10).map(_.toString).toList ::: "Jack" :: "Queen" :: "King" :: Nil
}
